### DISCUSSION TITLE
poor dogbolt performance

### DISCUSSION THREAD

**Discussion by psifertex** [NationalSecurityAgency/ghidra#5230] [Q&A]
Just wanted to reach out and ask if there's a way we could improve the performance of Ghidra on [dogbolt.org](https://dogbolt.org).

You can see how we currently automate decompilation here:

https://github.com/decompiler-explorer/decompiler-explorer/blob/master/runners/decompiler/decompile_ghidra.py#L28-L38
and:
https://github.com/decompiler-explorer/decompiler-explorer/blob/master/runners/decompiler/tools/ghidra/DecompilerExplorer.java

We're seeing significantly worse performance than I'd expect and I'm not sure whether there are any obvious improvements to the automation we can make and any advice from the official devs would be appreciated!

**Comment by astrelsky** [comment]
You could try disabling analyzers that may not be necessary. You should be able to see the analysis times in the log file. It should be fairly obvious which ones ~~have been neglected~~ could be disabled to improve analysis times and thus decompilation time. You'll need to play around with these settings to determine what works best for your use case.

For example (stolen from this [readme](https://github.com/clearbluejar/ghidrecomp)):
```
INFO  -----------------------------------------------------
    ASCII Strings                              0.343 secs
    Apply Data Archives                        0.548 secs
    Call Convention ID                         0.025 secs
    Call-Fixup Installer                       0.013 secs
    Create Address Tables                      0.039 secs
    Create Address Tables - One Time           0.076 secs
    Create Function                            0.232 secs
    Data Reference                             0.126 secs
    Decompiler Parameter ID                   10.939 secs
    Decompiler Switch Analysis                 0.442 secs
    Demangler Microsoft                        0.089 secs
    Disassemble                                0.007 secs
    Disassemble Entry Points                   1.959 secs
    Disassemble Entry Points - One Time        0.005 secs
    Embedded Media                             0.020 secs
    External Entry References                  0.000 secs
    Function ID                                0.805 secs
    Function Start Search                      0.038 secs
    Non-Returning Functions - Discovered       0.132 secs
    Non-Returning Functions - Known            0.012 secs
    PDB Universal                              2.810 secs
    Reference                                  0.104 secs
    Scalar Operand References                  0.601 secs
    Shared Return Calls                        0.219 secs
    Stack                                      5.164 secs
    Subroutine References                      0.113 secs
    Subroutine References - One Time           0.026 secs
    Windows x86 PE Exception Handling          0.027 secs
    Windows x86 PE RTTI Analyzer               0.005 secs
    Windows x86 Thread Environment Block (TEB) Analyzer     0.038 secs
    WindowsResourceReference                   0.124 secs
    x86 Constant Reference Analyzer            5.755 secs
-----------------------------------------------------
     Total Time   30 secs
-----------------------------------------------------
```

I did wait on responding with this hoping one of the actual devs would chime in.

**Reply by psifertex** [reply]
Thanks -- yeah, I was hoping they'd be able to help out as well, maybe they still will. 

Unfortunately, I can't easily determine in advance what to enable or what not to because users can upload any type of executable file and I don't know which pass is important for good decompilation quality of what file formats. The whole point is to showcase the best decompilation of every tool for comparison but right now the performance of Ghidra makes it very unfair which is not intended. In my local tests Ghidra is slower, but not _that_ much slower so I also wonder if there is some other hardware differences in the cloud infrastructure we're using that are an additional unnecessary performance tax (to be clear, all nodes run the same hardware, but it just appears Ghidra performs worse on them than the other tools do relative to standard desktop/laptop configurations).

**Reply by astrelsky** [reply]
What does the hardware look like? Specifically the os and architecture.

Where did the Ghidra installation come from?

**Reply by psifertex** [reply]
Ghidra download is from the latest stable release from the official github. 
Host is a digitalocean droplet running Ubuntu 20.04 w/ 8gb ram, 8gb of swap, 4 vcpu

**Reply by astrelsky** [reply]
I think that rules out anything really simple. I'm out of ideas for now.

**Comment by ryanmkurtz** [comment]
I just tried this out and for some reason the import of my binary [(link)](https://github.com/NationalSecurityAgency/ghidra/files/7448875/iOS15_example.zip) failed on dogbolt (see below output).

I then downloaded the Ghidra 10.2.2 release and your script, and ran

```
time ./analyzeHeadless /Users/ryan/ghidra_projects Test -import ~/samples/nslogtest2 -overwrite -postScript DecompilerExplorer.java out.c
```

and it succeeded.  I noted the major runtime discrepancies, but I guess first we need to figure out why the failure occurred.

```
INFO  Using log config file: jar:file:/home/decompiler_user/ghidra/Ghidra/Framework/Generic/lib/Generic.jar!/generic.log4j.xml (LoggingInitialization)  
INFO  Using log file: /home/decompiler_user/.ghidra/.ghidra_10.2.2_PUBLIC/application.log (LoggingInitialization)  
INFO  Loading user preferences: /home/decompiler_user/.ghidra/.ghidra_10.2.2_PUBLIC/preferences (Preferences)  
INFO  Class search complete (3992 ms) (ClassSearcher)  
INFO  Initializing SSL Context (SSLContextInitializer)  
INFO  Initializing Random Number Generator... (SecureRandomFactory)  
INFO  Random Number Generator initialization complete: NativePRNGNonBlocking (SecureRandomFactory)  
INFO  Trust manager disabled, cacerts have not been set (ApplicationTrustManagerFactory)  
INFO  HEADLESS Script Paths:
    /home/decompiler_user/ghidra/Ghidra/Debug/Debugger-agent-dbgmodel-traceloader/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Features/FunctionID/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Features/Decompiler/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Features/FileFormats/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Processors/Atmel/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Features/VersionTracking/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Processors/DATA/ghidra_scripts
    /home/decompiler_user
    /home/decompiler_user/ghidra/Ghidra/Debug/Debugger/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Processors/PIC/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Features/BytePatterns/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Features/PDB/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Features/GnuDemangler/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Features/MicrosoftCodeAnalyzer/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Features/Python/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Debug/Debugger-agent-frida/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Processors/8051/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Features/Base/ghidra_scripts
    /home/decompiler_user/ghidra/Ghidra/Processors/JVM/ghidra_scripts (HeadlessAnalyzer)  
INFO  HEADLESS: execution starts (HeadlessAnalyzer)  
INFO  Creating project: /tmp/tmp_dieyv9r/tmp44vdlesv/temp (HeadlessAnalyzer)  
INFO  Creating project: /tmp/tmp_dieyv9r/tmp44vdlesv/temp (DefaultProject)  
INFO  REPORT: Processing input files:  (HeadlessAnalyzer)  
INFO       project: /tmp/tmp_dieyv9r/tmp44vdlesv/temp (HeadlessAnalyzer)  
INFO  IMPORTING: /tmp/tmp_dieyv9r/tmptku0192r (HeadlessAnalyzer)  
INFO  No load spec found for import file: /tmp/tmp_dieyv9r/tmptku0192r (AutoImporter)  
ERROR The AutoImporter could not successfully load /tmp/tmp_dieyv9r/tmptku0192r with the provided import parameters. Please ensure that any specified processor/cspec arguments are compatible with the loader that is used during import and try again. (HeadlessAnalyzer)  
ERROR REPORT: Import failed for file: /tmp/tmp_dieyv9r/tmptku0192r (HeadlessAnalyzer)  

openjdk version "17.0.4" 2022-07-19
OpenJDK Runtime Environment (build 17.0.4+8-Debian-1deb11u1)
OpenJDK 64-Bit Server VM (build 17.0.4+8-Debian-1deb11u1, mixed mode)
```

**Reply by ryanmkurtz** [reply]
I just uploaded a second binary and it failed with a similar error.

**Reply by ryanmkurtz** [reply]
I tried a different binary with other decompilers enabled and got

<img width="1385" alt="image" src="https://user-images.githubusercontent.com/5378554/234889348-0c8b6a8c-0158-40ef-bfb0-21d0d6f1c771.png">

The Ghidra error was instant, which was odd since the other programs were still allowed to crank away until they finished.

**Reply by psifertex** [reply]
Let me double check the logs more closely but one possibility is that if a binary has previously errored on that exact version the result can be cached. If that's a unique binary unlikely to have been uploaded before that's not likely the cause though. 

If you look in the HTML source you'll see there's a refresh icon we disable by default -- if you turn that back on you can use that to force re-analysis.

**Reply by negasora** [reply]
The refresh button is shown to the right of the decompiler name

**Comment by ryanmkurtz** [comment]
Making a new thread for the performance issue.  What kind of debugging have you tried to isolate the issue?  For example, here are some things I'd try:

* Don't run a postScript
* Run a postScript that doesn't do anything
* Add the `-noanalysis` flag

Seeing how these things affect performance will help isolate the issue.
